pkg/auth/requests/service_account.go

```diff
@@ -0,0 +1,161 @@
+	restConfigGetter          restConfigGetter
+	authClientCreator         authClientCreator
+	clusterProxyConfigsGetter controllers.ClusterProxyConfigCache
+}
+
+// NewServiceAccountAuth creates a new instance of ServiceAccountAuth.
+func NewServiceAccountAuth(
+	scaledContext *config.ScaledContext,
+	restConfigGetter restConfigGetter,
+) auth.Authenticator {
+	return &ServiceAccountAuth{
+		scaledContext:    scaledContext,
+		clusterLister:    scaledContext.Management.Clusters("").Controller().Lister(),
+		secretLister:     scaledContext.Core.Secrets("").Controller().Lister(),
+		restConfigGetter: restConfigGetter,
+		authClientCreator: func(config *rest.Config) (authv1.AuthenticationV1Interface, error) {
+			return authv1.NewForConfig(config)
+		},
+		clusterProxyConfigsGetter: scaledContext.Wrangler.Mgmt.ClusterProxyConfig().Cache(),
+	}
+}
+
+// Authenticate the request using the downstream service account's JWT.
+func (t *ServiceAccountAuth) Authenticate(req *http.Request) (user.Info, bool, error) {
+	// Basic checks to see if we even need to proceed
+	info, hasAuth := request.UserFrom(req.Context())
+	if info.GetName() != "system:cattle:error" {
+		return info, hasAuth, nil
+	}
+	clusterID := mux.Vars(req)["clusterID"]
+	if clusterID == "" {
+		return info, hasAuth, fmt.Errorf("no clusterID found in request")
+	}
+	// Check the cluster setting value to determine whether we will continue the auth process
+	settings, err := t.clusterProxyConfigsGetter.Get(clusterID, settingsObjectName)
+	if err != nil {
+		logrus.Warnf("rejecting downstream proxy request for %s, unable to fetch ClusterProxySettings object for cluster", req.URL.Path)
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	if !settings.Enabled {
+		logrus.Warnf("rejecting downstream proxy request for %s, current setting is enabled: %v", req.URL.Path, settings.Enabled)
+		return &user.DefaultInfo{}, false, nil
+	}
+
+	// See if the token is a JWT.
+	rawToken := tokens.GetTokenAuthFromRequest(req)
+
+	jwtParser := jwtv4.Parser{}
+	claims := jwtv4.RegisteredClaims{}
+	_, _, err = jwtParser.ParseUnverified(rawToken, &claims)
+	if err != nil {
+		logrus.Debugf("saauth: error parsing JWT: %v", err)
+		return info, false, err
+	}
+
+	if !strings.HasPrefix(claims.Subject, serviceAccountSubjectPrefix) {
+		logrus.Debugf("saauth: JWT sub is not a service account: %v", err)
+		return info, false, nil
+	}
+
+	if isTokenExpired(claims.ExpiresAt) {
+		logrus.Debugf("saauth: Service Account JWT is expired. Expiration time was: %v", claims.ExpiresAt)
+		return info, false, nil
+	}
+
+	cluster, err := t.clusterLister.Get("", clusterID)
+	if err != nil {
+		logrus.Debugf("saauth: error getting cluster: %v", err)
+		return info, false, err
+	}
+
+	// Get rest config for the cluster and instantiate an authentication client.
+	kubeConfig, err := t.restConfigGetter(cluster, t.scaledContext, t.secretLister)
+	if kubeConfig == nil || err != nil {
+		return info, false, err
```

Comment:
I would not recommend returning this error to the end user - this could expose internal details of the Rancher app that the end user doesn't need to see. I would recommend returning a more generic internal error message (such as "failed to get kubeconfig when validating token for downstream cluster") and logging the full error.